charms/kserve-web-app/src/charm.py

```diff
@@ -13,43 +13,84 @@
 """
 
 import logging
+import traceback
 
+from charms.observability_libs.v0.kubernetes_service_patch import KubernetesServicePatch
 from ops.charm import CharmBase
-from ops.framework import StoredState
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, BlockedStatus
+from ops.pebble import ChangeError, Layer
 
-logger = logging.getLogger(__name__)
+log = logging.getLogger(__name__)
 
 
 class KServeWebAppCharm(CharmBase):
     """Charm the service."""
 
-
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_kserve_web_app_ready)
+        self.framework.observe(self.on.kserve_web_app_pebble_ready, self._on_kserve_web_app_ready)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
+
+        self._container_name = "kserve-web-app"
+        self.container = self.unit.get_container(self._container_name)
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [(self._container_name, self.model.config["port"])]
+        )
+
+    @property
+    def _pebble_layer(self):
+        """Return the Pebble layer for the workload."""
+        return Layer(
+            {
+                "services": {
+                    "kserve-web-app": {
+                        "override": "replace",
+                        "summary": "KServe Web App",
+                        "command": "gunicorn -w 3 --bind 0.0.0.0:5000 --access-logfile - entrypoint:app",
+                        "startup": "enabled",
+                        "environment": {
+                            "USERID_HEADER": "kubeflow-userid",
+                            "APP_PREFIX": "/kserve-endpoints",
+                            "APP_DISABLE_AUTH": True,
+                        },
+                    }
+                }
+            }
+        )
 
     def _on_kserve_web_app_ready(self, event):
         """Define and start a workload using the Pebble API.
 
         Learn more about Pebble layers at https://github.com/canonical/pebble
         """
-        pass
+        self._update_layer()
+        self.unit.status = ActiveStatus()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+    def _on_config_changed(self, event):
+        """Config Changed event handler.
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
-
-        Learn more about config at https://juju.is/docs/sdk/config
+        We skip the handler as everything is done in init
```

Comment:
Is it mandatory to have this?  Or is it being included here to signal to a reader "we know it exists, we just don't need it"?

Reply:
I'm not sure how else I would observe the config changed event without it doing something unnecessary, but more than open to alternatives.

Reply:
why do you need to observe it in the first place?

Reply:
So that the charm calls the service patch library in init when we change the port number.

Reply:
Hum... Are you suggesting that, if you don't register an observer to config-changed, then the charm's `__init__` will not be executed when the unit receives a config-changed event?

Reply:
The charm's `__init__` is guaranteed to run, regardless of the event being fired, and regardless of whether the charm or any of its libs are observing that specific event.

- unit receives an event
- charm is instantiated (`__init__` runs)
- any and all observers are executed (none if this is a config-changed event)

Reply:
Oh I didn't know that, in that case I don't need to observe the event. Thanks!